src/Multicall3.sol

```diff
@@ -0,0 +1,187 @@
+    function tryBlockAndAggregate(bool requireSuccess, Call[] calldata calls) public payable returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = tryAggregate(requireSuccess, calls);
+    }
+
+    /// @notice Backwards-compatible with Multicall2
+    /// @notice Aggregate calls and allow failures using tryAggregate
+    /// @param calls An array of Call structs
+    /// @return blockNumber The block number where the calls were executed
+    /// @return blockHash The hash of the block where the calls were executed
+    /// @return returnData An array of Result structs
+    function blockAndAggregate(Call[] calldata calls) public payable returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        (blockNumber, blockHash, returnData) = tryBlockAndAggregate(true, calls);
+    }
+
+    /// @notice Aggregate calls, ensuring each returns success if required
+    /// @param calls An array of Call3 structs
+    /// @return blockNumber The block number where the calls were executed
+    /// @return blockHash The hash of the block where the calls were executed
+    /// @return returnData An array of Result structs
+    function aggregate3(Call3[] calldata calls) public payable returns (uint256 blockNumber, bytes32 blockHash, Result[] memory returnData) {
+        blockNumber = block.number;
+        blockHash = blockhash(block.number);
+        returnData = new Result[](calls.length);
+        uint256 length = calls.length;
+        Call3 calldata call;
+        for (uint256 i = 0; i < length;) {
+            Result memory result = returnData[i];
+            call = calls[i];
+            (result.success, result.returnData) = call.target.call(call.callData);
+            require(call.allowFailure || result.success, "Multicall3: aggregate3 failed");
+            unchecked { ++i; }
+        }
+    }
+
+    /// @notice Aggregate calls with a msg value
+    /// @notice Reverts if msg.value is less than the sum of the call values
+    /// @param calls An array of Call3Value structs
+    /// @return returnData An array of Result structs
+    function aggregate3Value(Call3Value[] calldata calls) public payable returns (Result[] memory returnData) {
+        uint256 valAccumulator;
+        returnData = new Result[](calls.length);
+        uint256 length = calls.length;
+        Call3Value calldata call;
+        for (uint256 i = 0; i < length;) {
+            Result memory result = returnData[i];
+            call = calls[i];
+            // Humanity will be a Type V Kardashev Civilization before this overflows - andreas
+            // ~ 10^25 Wei in existence << ~ 10^76 size uint fits in a uint256
+            unchecked { valAccumulator += call.value; }
+            (result.success, result.returnData) = call.target.call{value: call.value}(call.callData);
```

Comment:
```suggestion
            uint256 val = call.value;
            // Humanity will be a Type V Kardashev Civilization before this overflows - andreas
            // ~ 10^25 Wei in existence << ~ 10^76 size uint fits in a uint256
            unchecked { valAccumulator += val; }
            (result.success, result.returnData) = call.target.call{value: val}(call.callData);
```

```diff
diff --git a/.gas-snapshot b/.gas-snapshot
index def28f3..9fb01fb 100644
--- a/.gas-snapshot
+++ b/.gas-snapshot
@@ -15,8 +15,8 @@ Multicall2Test:testTryBlockAndAggregateUnsuccessful() (gas: 10382)
 Multicall3Test:testAggregate3Unsuccessful() (gas: 10279)
-Multicall3Test:testAggregate3Value() (gas: 34757)
-Multicall3Test:testAggregate3ValueUnsuccessful() (gas: 62666)
+Multicall3Test:testAggregate3Value() (gas: 34736)
+Multicall3Test:testAggregate3ValueUnsuccessful() (gas: 62608)
 Multicall3Test:testAggregation() (gas: 5883)
```

Reply:
Great catch! Added in #6 ! If you have any other suggestions feel free to fork and pr into this branch!

Reply:
done in d2320fb